taichi/llvm/llvm_program.cpp

```diff
@@ -473,6 +473,13 @@ void LlvmProgramImpl::finalize() {
     cuda_device()->dealloc_memory(preallocated_device_buffer_alloc);
   }
 #endif
+  for (auto &alloc : ndarray_allocs_) {
+    if (config->arch == Arch::cuda) {
+      cuda_device()->dealloc_memory(alloc);
```

Comment:
Here and below: it seems that Device is the only difference in many of the function definitions. Consider de-duplicating the code w/ sth like this:

```cpp
auto * device = (arch == cuda) ? cuda_device() : cpu_device();
// use `device`
```